private/rules/pin_dependencies.bzl

```diff
@@ -0,0 +1,113 @@
+load("//:coursier.bzl", "compute_dependency_inputs_signature")
+
+_TEMPLATE = """#!/usr/bin/env bash
+
+{resolver_cmd} --argsfile {config} --resolver {resolver} --input_hash '{input_hash}' --output {output}
+"""
+
+def _stringify_exclusions(exclusions):
+    return ["%s:%s" % (e["group"], e["artifact"]) for e in exclusions]
+
+def _pin_dependencies_impl(ctx):
+    # Repositories should be simple strings
+    repos = []
+    for r in ctx.attr.repositories:
+        repo = json.decode(r)
+        if repo.get("credentials"):
+            fail("Please use your .netrc file for access credentials to repos: ", repo["repo_url"])
+        repos.append(repo["repo_url"])
+
+    exclusions = []
+    for e in ctx.attr.excluded_artifacts:
+        exclusion = json.decode(e)
+        exclusions.append("%s:%s" % (exclusion["group"], exclusion["artifact"]))
+
+    artifacts = []
+    for a in ctx.attr.artifacts:
+        artifact = json.decode(a)
+        artifact["exclusions"] = _stringify_exclusions(artifact.get("exclusions", []))
+        artifacts.append(artifact)
+
+    boms = []
+    for b in ctx.attr.boms:
+        bom = json.decode(b)
+        bom["exclusions"] = _stringify_exclusions(bom.get("exclusions", []))
+        boms.append(bom)
+
+    # Prep the external config file
+    config = {
+        "repositories": repos,
+        "artifacts": artifacts,
+        "boms": [json.decode(b) for b in ctx.attr.boms],
+        "globalExclusions": exclusions,
+        "fetchSources": ctx.attr.fetch_sources,
+        "fetchJavadoc": ctx.attr.fetch_javadocs,
+    }
+
+    config_file = ctx.actions.declare_file("%s-config.json" % ctx.label.name)
+    ctx.actions.write(
+        config_file,
+        content = json.encode_indent(config, indent = "  "),
+    )
+
+    input_hash = compute_dependency_inputs_signature(
+        ctx.attr.artifacts,
```

Comment:
need to add `ctx.attr.boms,` as the first argument so `bazel run @unpinned_maven//:pin` doesn't fail with
```
File "/Users/cheister/.cache/bazel/879c5fa76919d0bed49168a2e457a904/external/rules_jvm_external/coursier.bzl", line 324, column 5, in compute_dependency_inputs_signature
		def compute_dependency_inputs_signature(boms, artifacts, repositories, excluded_artifacts):
Error: compute_dependency_inputs_signature() missing 1 required positional argument: excluded_artifacts
```